lib/vnet/user_process_windows.go

```diff
@@ -42,36 +46,124 @@ func runPlatformUserProcess(ctx context.Context, config *UserProcessConfig) (pm
-		<-processCtx.Done()
-		return trace.Wrap(listener.Close())
-	})
 	pm.AddCriticalBackgroundTask("admin process", func() error {
+		log.InfoContext(processCtx, "Starting Windows service")
 		return trace.Wrap(runService(processCtx, &windowsAdminProcessConfig{
 			clientApplicationServiceAddr: listener.Addr().String(),
+			serviceCredentialPath:        credDir,
 		}))
 	})
 	pm.AddCriticalBackgroundTask("gRPC service", func() error {
-		log.InfoContext(processCtx, "Starting gRPC service", "addr", listener.Addr().String())
-		// TODO(nklaassen): add mTLS credentials for client application service.
-		grpcServer := grpc.NewServer(
-			grpc.Creds(insecure.NewCredentials()),
-			grpc.UnaryInterceptor(interceptors.GRPCServerUnaryErrorInterceptor),
-			grpc.StreamInterceptor(interceptors.GRPCServerStreamErrorInterceptor),
-		)
-		clock := clockwork.NewRealClock()
-		appProvider := newLocalAppProvider(config.ClientApplication, clock)
-		svc := newClientApplicationService(appProvider)
-		vnetv1.RegisterClientApplicationServiceServer(grpcServer, svc)
-		if err := grpcServer.Serve(listener); err != nil {
-			return trace.Wrap(err, "serving VNet user process gRPC service")
-		}
+		log.InfoContext(processCtx, "Starting gRPC service",
+			"addr", listener.Addr().String())
+		return trace.Wrap(grpcServer.Serve(listener),
+			"serving VNet user process gRPC service")
+	})
+	pm.AddCriticalBackgroundTask("gRPC server closer", func() error {
+		// grpcServer.Serve does not stop on its own when processCtx is done, so
+		// this task waits for processCtx and then explicitly stops grpcServer.
+		<-processCtx.Done()
+		grpcServer.GracefulStop()
 		return nil
 	})
 	return pm, nil
 }
+
+// secureCredDir sets ACLs so that the current user can write credentials files
+// to dir but cannot read them. Only the LocalSystem account is allowed to read
+// and delete the files.
+func secureCredDir(dir string) error {
+	u, err := user.Current()
+	if err != nil {
+		return trace.Wrap(err, "getting current OS user")
+	}
+	// Uid is documented to be the user's SID on Windows.
+	userSID := u.Uid
+	// S-1-5-18 is the well-known SID (security identifier) for the LocalSystem
+	// service account, which the VNet windows service runs as.
+	// * https://learn.microsoft.com/en-us/windows/win32/secauthz/well-known-sids
+	const serviceSID = "S-1-5-18"
+	// O: Set Owner as current user (Windows won't allow LocalSystem to be the owner)
+	// G: Set Primary Group as current user
+	// D: Discretionary ACL
+	//   Grant GA (Generic All) to LocalSystem
+	//   Grant GW (Generic Write) to the current user so we can add files
+	//   Deny GR (Generic Read) for the current user so other processes can't read the credentials
+	//   OICI (Object Inherit, Container Inherit) propagates permissions to files/folders
+	sdString := fmt.Sprintf("O:%[1]sG:%[1]sD:"+
+		"(A;OICI;GA;;;%[2]s)"+
+		"(A;OICI;GW;;;%[1]s)"+
+		"(D;OICI;GR;;;%[1]s)",
+		userSID, serviceSID,
+	)
```

Comment:
I'm not an expert in Windows ACLs; does the above mean that apart from propagating the ACLs to subfolders, the secure folder will also inherit ACLs from its parent? Also: does it mean that users other than the current and LocalSystem won't be allowed to do anything with these files?

Reply:
> I'm not an expert in Windows ACLs

nobody is :)

it may or may not inherit from the parent, depending on the settings on the parent. I switched to explicitly deny Generic Read for Everyone instead of just the current user

> Also: does it mean that users other than the current and LocalSystem won't be allowed to do anything with these files?

depends on the default settings for wherever os.MkdirTemp puts this, but i think denying Generic Read for everyone should make this pretty safe